src/modules/dotnet.rs

```diff
@@ -0,0 +1,308 @@
+        return None;
+    }
+
+    let mut module = context.new_module("dotnet");
+
+    // Internally, this module uses its own mechanism for version detection.
+    // Typically it is twice as fast as running `dotnet --version`.
+    let enable_heuristic = module.config_value_bool("heuristic").unwrap_or(true);
+    let version = if enable_heuristic {
+        let repo_root = context
+            .get_repo()
+            .ok()
+            .and_then(|r| r.root.as_ref().map(|buf| buf.as_path()));
+        estimate_dotnet_version(&dotnet_files, &context.current_dir, repo_root)?
+    } else {
+        get_version_from_cli()?
+    };
+
+    let module_style = module
+        .config_value_style("style")
+        .unwrap_or_else(|| Color::Blue.bold());
+
+    module.set_style(module_style);
+    module.new_segment("symbol", DOTNET_SYMBOL);
+    module.new_segment("version", &version);
+
+    Some(module)
+}
+
+fn estimate_dotnet_version<'a>(
+    files: &[DotNetFile<'a>],
+    current_dir: &Path,
+    repo_root: Option<&Path>,
+) -> Option<Version> {
+    let get_file_of_type = |t: FileType| files.iter().find(|f| f.file_type == t);
+
+    // Check for a global.json or a solution file first, but otherwise take the first file
+    let relevant_file = get_file_of_type(FileType::GlobalJson)
+        .or_else(|| get_file_of_type(FileType::SolutionFile))
+        .or_else(|| files.iter().next())?;
+
+    match relevant_file.file_type {
+        FileType::GlobalJson => {
+            get_pinned_sdk_version_from_file(relevant_file.path).or_else(get_latest_sdk_from_cli)
+        }
+        FileType::SolutionFile => {
+            // With this heuristic, we'll assume that a "global.json" won't
+            // be found in any directory above the solution file.
+            get_latest_sdk_from_cli()
+        }
+        _ => {
+            // If we see a dotnet project, we'll check a small number of neighboring
+            // directories to see if we can find a global.json. Otherwise, assume the
+            // latest SDK is in use.
+            try_find_nearby_global_json(current_dir, repo_root).or_else(get_latest_sdk_from_cli)
+        }
+    }
+}
+
+fn try_find_nearby_global_json(current_dir: &Path, repo_root: Option<&Path>) -> Option<Version> {
+    let current_dir_is_repo_root = repo_root.map(|r| r == current_dir).unwrap_or(false);
+    let parent_dir = if current_dir_is_repo_root {
+        // Don't scan the parent directory if it's above the root of a git repository
+        None
+    } else {
+        current_dir.parent()
+    };
+
+    // Check the parent directory, or otherwise the repository root, for a global.json
+    let mut check_dirs = parent_dir
+        .iter()
+        .chain(repo_root.iter())
+        .copied() // Copies the reference, not the Path itself
+        .collect::<Vec<&Path>>();
+
+    // The parent directory and repository root may be the same directory,
+    // so avoid checking it twice.
+    check_dirs.dedup();
```

Comment:
Is there a guarantee that the parent directory and repo root will be consecutive in the vector? According to the doc, `dedup()` only deduplicates consecutive elements.

Reply:
Yep, if you check above, the vector will have at most two elements.